keras_nlp/models/t5/t5_backbone.py

```diff
@@ -0,0 +1,846 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""T5 backbone model."""
+
+import copy
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.compiler.tf2xla.python.xla import dynamic_slice
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.models.backbone import Backbone
+from keras_nlp.models.t5.t5_presets import backbone_presets
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.T5Backbone")
+class T5Backbone(Backbone):
+    def __init__(
+        self,
+        vocabulary_size,
+        output_dim,
+        hidden_dim,
+        key_value_dim,
+        num_blocks,
+        num_heads,
+        use_gated_activation,
+        activation,
+        dropout,
+        layer_norm_epsilon,
+        **kwargs,
+    ):
+        shared_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=output_dim,
+            embeddings_initializer=keras.initializers.TruncatedNormal(1.0),
+            name="shared_embedding",
+        )
+        encoder = T5MainLayer(
+            is_decoder=False,
+            use_gated_activation=use_gated_activation,
+            output_dim=output_dim,
+            hidden_dim=hidden_dim,
+            dropout=dropout,
+            activation=activation,
+            layer_norm_epsilon=layer_norm_epsilon,
+            key_value_dim=key_value_dim,
+            num_heads=num_heads,
+            num_blocks=num_blocks,
+            token_embeddings=shared_embedding,
+            name="encoder",
+        )
+        decoder = T5MainLayer(
+            is_decoder=True,
+            use_gated_activation=use_gated_activation,
+            output_dim=output_dim,
+            hidden_dim=hidden_dim,
+            dropout=dropout,
+            activation=activation,
+            layer_norm_epsilon=layer_norm_epsilon,
+            key_value_dim=key_value_dim,
+            num_heads=num_heads,
+            num_blocks=num_blocks,
+            token_embeddings=shared_embedding,
+            name="decoder",
+        )
+        token_ids = keras.Input(shape=(None,), dtype="int32", name="token_ids")
+        decoder_token_ids = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_token_ids"
+        )
+        encoder_outputs = encoder(token_ids)
+        decoder_outputs = decoder(
+            decoder_token_ids,
+            encoder_hidden_states=encoder_outputs,
+        )
+        super().__init__(
```

Comment:
I think we should probably just follow this input and output for now...
https://github.com/keras-team/keras-nlp/blob/4c94a0d68872a72ed3e27d3f93a3d7eec3ddc0ba/keras_nlp/models/bart/bart_backbone.py#L217-L226

Padding mask vs whole attention mask is unclear to me, but our convention is to take a padding mask on every model, so let's do it here.

And it does seem useful for the backbone to return both the encoder and decoder outputs? That gives you the full "dense representation" at the backbone level.